File: LatitudeSolver.swift
import UIKit

/**
 iOS Enlem Çözücüsü
 
 Polaris'in piksel konumundan enlemi hesaplar.
 */
struct LatitudeResult {
    let latitude: Double
    let lowerBound: Double
    let upperBound: Double
    let errorMargin: Double
    let altitude: Double
}

class LatitudeSolver {
    
    func calculateLatitude(
        polariPixelY: CGFloat,
        imageHeight: CGFloat,
        verticalFov: Double
    ) -> LatitudeResult {
        // Merkez noktasından sapma
        let centerY = imageHeight / 2
        let pixelOffset = centerY - polariPixelY
        
        // Dereceye çevir
        let degreesPerPixel = verticalFov / Double(imageHeight)
        let altitude = Double(pixelOffset) * degreesPerPixel
        
        // Polaris yüksekliği ≈ Enlem
        let latitude = altitude
        
        // Hata hesabı
        let fovUncertainty: Double = 2.0
        let calibrationError: Double = 1.0
        
        let latWithLowFov = Double(centerY - polariPixelY) * ((verticalFov - fovUncertainty) / Double(imageHeight))
        let latWithHighFov = Double(centerY - polariPixelY) * ((verticalFov + fovUncertainty) / Double(imageHeight))
        
        let fovError = max(abs(latWithLowFov - latitude), abs(latWithHighFov - latitude))
        let totalError = sqrt(fovError * fovError + calibrationError * calibrationError)
        
        return LatitudeResult(
            latitude: round(latitude * 100) / 100,
            lowerBound: round((latitude - totalError) * 100) / 100,
            upperBound: round((latitude + totalError) * 100) / 100,
            errorMargin: round(totalError * 100) / 100,
            altitude: round(altitude * 100) / 100
        )
    }
    
    func pixelToAngle(pixelOffset: CGFloat, imageHeight: CGFloat, verticalFov: Double) -> Double {
        let degreesPerPixel = verticalFov / Double(imageHeight)
        return Double(pixelOffset) * degreesPerPixel
    }
}

File: LatitudeSolver.kt
package com.kutup.navigasyon

import kotlin.math.sqrt

/**
 * Enlem Çözücüsü
 * 
 * Polaris'in piksel konumundan enlemi hesaplar.
 * Formül: Polaris'in ufuk üstündeki yüksekliği = Bulunduğunuz Enlem
 */
data class LatitudeResult(
    val latitude: Float,
    val lowerBound: Float,
    val upperBound: Float,
    val errorMargin: Float,
    val altitude: Float
)

class LatitudeSolver {
    
    fun calculateLatitude(
        polariPixelY: Float,
        imageHeight: Int,
        verticalFov: Float
    ): LatitudeResult {
        // Merkez noktasından sapma
        val centerY = imageHeight / 2f
        val pixelOffset = centerY - polariPixelY
        
        // Dereceye çevir
        val degreesPerPixel = verticalFov / imageHeight
        val altitude = pixelOffset * degreesPerPixel
        
        // Polaris yüksekliği ≈ Enlem
        val latitude = altitude
        
        // Hata hesabı
        val fovUncertainty = 2f  // FOV ölçüm hatasında ±2° varsay
        val calibrationError = 1f  // Kalibrasyon hatası ±1°
        
        val latWithLowFov = (centerY - polariPixelY) * ((verticalFov - fovUncertainty) / imageHeight)
        val latWithHighFov = (centerY - polariPixelY) * ((verticalFov + fovUncertainty) / imageHeight)
        
        val fovError = maxOf(
            kotlin.math.abs(latWithLowFov - latitude),
            kotlin.math.abs(latWithHighFov - latitude)
        )
        
        val totalError = sqrt(fovError * fovError + calibrationError * calibrationError)
        
        return LatitudeResult(
            latitude = round(latitude, 2),
            lowerBound = round(latitude - totalError, 2),
            upperBound = round(latitude + totalError, 2),
            errorMargin = round(totalError, 2),
            altitude = round(altitude, 2)
        )
    }
    
    fun pixelToAngle(pixelOffset: Float, imageHeight: Int, verticalFov: Float): Float {
        val degreesPerPixel = verticalFov / imageHeight
        return pixelOffset * degreesPerPixel
    }
    
    private fun round(value: Float, decimals: Int): Float {
        val multiplier = kotlin.math.pow(10f, decimals.toFloat()).toInt()
        return (value * multiplier).toInt() / multiplier.toFloat()
    }
}
